### PR TITLE
Tests in meson

### PR DESCRIPTION
**Pull request by rcombs**
Adds tests for meson builds in CI.

### CODE REVIEW

File: .github/workflows/meson.yml
Comment (kasper93) [inline .github/workflows/meson.yml:104]: Mixing mingw build with msvc one defeats the purpose, no? (and is generally unsafe, if public headers are not carefully vetted to work) All those libraries are available in meson's wrap db which make it very easy to build when needed.

File: compare/meson.build
Comment (kasper93) [inline compare/meson.build:18]: I would suggest using native binaries to run tests. It is more portable. As it is defined currently, meson doesn't know that the binary is executed, only path is passed to the script. Therefore it won't add varaibles to environment like PATH, ASAN_OPTIONS and most imprtantly when cross compiling it wouldn't know to use exe_wrapper (like wine or qemu) if needed. All this makes running tests portable a nightmare, while it would be already handled by Meson.

The easiest and canonical way would be to loop the files in meson and declare test for each of them, instead of using `run_all.sh` helper.

File: fuzz/fuzz.c
Comment (kasper93) [inline fuzz/fuzz.c:175]: Could it be in some common header? Also `<io.h>` and `<unistd.h>` could be included from common header.

Also, that's not correct for mingw build. This should be done only for MSVC. And using larger type is fine, makes easier for downstream compatibility.

```suggestion
#ifdef _MSC_VER
typedef long long ssize_t;
#endif
```

File: fuzz/fuzz.c
Comment (kasper93) [inline fuzz/fuzz.c:188]: You can easily defined those helper nacros in MSVC case in some common header.

``` c
#ifdef _MSC_VER
#ifndef STDIN_FILENO
#define STDIN_FILENO 0
#endif
#ifndef STDOUT_FILENO
#define STDOUT_FILENO 1
#endif
#ifndef STDERR_FILENO
#define STDERR_FILENO 2
#endif
#endif
```

File: fuzz/writeout.c
Comment (kasper93) [inline fuzz/writeout.c:190]: Unrelated note, but for fuzzing on Linux it is better to use `memfd_create` in such cases. See for example https://github.com/mpv-player/mpv/blob/baf528069a584c04686262ef6f76e2c9232adba3/fuzzers/fuzzer_load.c#L45-L58

File: fuzz/fuzz.c
Comment (rcombs) [reply fuzz/fuzz.c:175]: Win32 defines `ssize_t` as an unsigned type.

File: compare/meson.build
Comment (TheOneric) [inline compare/meson.build:7]: `libass_test` is only namespaced because meson reserves the `test` name; does the same apply here?

File: checkasm/meson.build
Comment (TheOneric) [inline checkasm/meson.build:29]: incorrect linkage; this will break on dynamic-only builds on Windows since checkasm relies on accessing private libass functions *(only on Windows because meson still lacks cross-platform symbol export control)*

File: compare/compare.c
Comment (TheOneric) [inline compare/compare.c:24]: Using libass internal functions means this can no longer be built with dynamic libass

Either avoid this by just not supporting `compare` with MSVC, or force static linkage like for `checkasm` i both autotools and meson

File: meson.build
Comment (TheOneric) [inline meson.build:124]: since it’s no longer used exclusively by `test`, this should be renamed

File: fuzz/writeout.c
Comment (TheOneric) [reply fuzz/writeout.c:190]: This code is not even compiled in for continuous fuzzing builds and `memfd_create` defeats the purpose of the writeout functionality

File: checkasm/meson.build
Comment (rcombs) [reply checkasm/meson.build:29]: Should I just disable checkasm and compare when building shared-only? (`compare` is already off by default)

File: compare/compare.c
Comment (astiob) [reply compare/compare.c:24]: > by just not supporting `compare` with MSVC

Don’t we need it for tests? Isn’t this (it being unsupported with MSVC) the current status quo that this PR is aiming to change?

Comment (TheOneric) [reply compare/compare.c:24]: We need it for full test coverage, yes. If the goal of the PR is to get the Windows threading implementation tested in CI, MinGW is sufficient if i’m not mistaken about it also using Win32-native API. If the goal is to get specifically MSVC tested in CI, then yeah, omitting `compare` support goes against that

File: checkasm/meson.build
Comment (TheOneric) [reply checkasm/meson.build:29]: In autotools we split the actual libass lib build into two stages *(building internal static-only lib; pushing this through the linker to get the final lib)* to be able to always provide a static lib for checkasm, so if you want to keep `ass_filesystem` in `compare`, you can use this too.
Given meson’s position as a secondary build system with limited support, just erroring out if a tool needing internal API is requested in a dynamic-only build is probably fine too.

File: meson.build
Comment (TheOneric) [inline meson.build:379]: just like every other option in the autotools and meson config setups, this should error out if it was explicitly requested but is unavailable instead of silently omitting it

File: meson.build
Comment (TheOneric) [inline meson.build:388]: same as with `compare`

File: compare/compare.c
Comment (astiob) [reply compare/compare.c:24]: Not sure about threads, but at the very least, MSVC builds use different atomics from MinGW.

**Comment by rcombs** [comment]
All build issues are now resolved; this now succeeds on all targets. I believe the meson feature-management issues are also resolved. This should be ready for merge once approved. I think #793 should be blocked on this, since running the full test suite with threading enabled on the MSVC build will be very useful.

I might also add a test pass to the muon build, but I think that can be scoped outside of this PR. The muon build might end up being our "threading-disabled" CI build in the medium-term (as it has bugs that prevent threading from playing nicely with asm).

File: meson.build
Comment (kasper93) [inline meson.build:380]: What's up with `.enabled()`? If it is intended to be a boolean option instead of feature, it should be defined as such in `meson_options.txt`.

File: meson.build
Comment (kasper93) [inline meson.build:394]: So to enable libass tests, user has to enable 4 different flags? Why? Could there be single a single flag? It's unclear to me, the user, how to test libass before deployment.

File: meson.build
Comment (kasper93) [inline meson.build:383]: Why this is disabled in shared build?

File: compare/image.c
Comment (kasper93) [inline compare/image.c:36]: Is this expected that indentation is different than in the rest of the file?

File: .github/workflows/meson.yml
Comment (kasper93) [inline .github/workflows/meson.yml:55]:
```suggestion
      ART_SAMPLES: ${{ github.workspace }}/ext_art-samples
```

File: .github/workflows/meson.yml
Comment (TheOneric) [reply .github/workflows/meson.yml:104]: Putting aside whether this mixing is in general safe or not, the build setup is more complex yet slower than wrapdb+sccache from #857. The latter brought down MSVC build times into the range of other Windows jobs, while the mixed setup here sticks out taking ~60% more time than the next slowest CI job.

Perhaps the argument can be made for there being some merit in having this setup in git history as a point of reference for what it takes to get a mixed setup working; but for further use sccache+wrapdb just seems more beneficial to me

File: meson.build
Comment (TheOneric) [reply meson.build:383]: Because it _requires_ access to libass-internal symbols which is not possible with a proper dynamic library.
The autotools build has a way to deal with this which can eventually be adapted into the meson setup, but for now just getting tests at all and restricting this is fine imo

Comment (kasper93) [reply meson.build:383]: If a tool or test depends on internal symbols, indeed it shouldn't like with libass directly. Meson has easy way to handle this case, without splitting into static libraries, `libass.extract_all_objects(recursive: true)`  or `libass.extract_objects(...)` for more fine-grained objects to link.

File: meson.build
Comment (rcombs) [reply meson.build:394]: This is targeting parity with the autoconf build; simplifying the flags would be reasonable, but is out-of-scope for this PR.

**Comment by rcombs** [comment]
I think this is in a good position to merge at this point. There are some suggestions to adjust the deps situation on windows, but I don't think they need to block this.

File: meson.build
Comment (kasper93) [reply meson.build:394]: Ok, clear. So how should I test libass? Is `-Dtest=enabled` good enough as integration tests or should I enable all other flags too?

File: compare/image.c
Comment (astiob) [inline compare/image.c:67]: If I understand correctly, this is a workaround for the MinGW-originating libpng using a different libc (and thus a different `FILE`) from libass on Windows?

Firstly, this deserves a comment, because this code looks like it’s completely redundant. Secondly, try changing the `msystem` in the GitHub CI matrix to `UCRT64`: assuming a new-enough MSVC, I think this should get you the right libc, making this code unnecessary. (Thirdly, of course, this is just another argument for not sourcing dependencies for an MSVC build from MSYS2.)

File: compare/meson.build
Comment (astiob) [inline compare/meson.build:18]: How does this work anyway? AFAICT the shell is set to PowerShell, but the executable is a Bourne shell script. Is MSYS2 automatically registered as a handler for `sh` files?

File: meson.build
Comment (astiob) [reply meson.build:394]: This repository doesn’t contain any integration tests; only the infrastructure.

These flags control which utilities are built. `checkasm` can be run to verify that our assembler code matches our C code and is run by `meson test`/`make check`. The others are special-purpose utilities that can be run by hand; in particular, `test` just takes an ASS file, renders it onto a blank canvas and saves it as a PNG.

Actual integration tests are in the dedicated libass-tests repository, which contains a shell script to run them using these utilities: `compare` for regression tests and `fuzz` for crash tests.

So if you want to run all tests, you must check out the libass-tests repo, build libass with `checkasm`, `compare` and `fuzz` (but not necessarily `test`) enabled, and either
* set `-Dart-samples=path/to/libass-tests` in Meson and run `meson test` (which, at the current state of this PR, will run the tests but only give a summary verdict for all of them together)
* or run `meson test` without defining `art-samples` (which will only run `checkasm`) and run libass-tests separately (which will let you see individual test results).

File: compare/meson.build
Comment (astiob) [reply compare/meson.build:18]: Further, looking at the CI output, it seems that this causes Meson to swallow the output of our test scripts and just print a single “OK” for the whole lot of tests. Which is certainly better than nothing—and at the end of the day we can run the tests manually to find the failing test—but not ideal. Although maybe it prints the output if the tests should fail?

Actually, does it even properly detect failures?

But anyway, on the other hand, the suggested solution seems to amount to a full rewrite of our test infrastructure in Meson, which also isn’t ideal. On the third hand, this might be easier than it sounds, because some of that code is too generalized/configurable for basic CI and could perhaps be simplified during this rewrite.

Comment (arch1t3cht) [reply compare/meson.build:18]: By default meson will redirect the output into `meson-logs`, but this can be configured with e.g. `-v`

Comment (kasper93) [reply compare/meson.build:18]: > Although maybe it prints the output if the tests should fail?

It logs everything in test-log and can be enable to stdout with verbose flag. Though if it passes shorter output generally make it easier to ready, when there is no need for more info.

> But anyway, on the other hand, the suggested solution seems to amount to a full rewrite of our test infrastructure in Meson

I don't see it that way, I had a glance at run_all.sh script. And it is essentiality aggregator of testcases. So what already Meson does and sure it would be duplicating some logic, so running script like that is fine, but as you noticed it bundles all tests into one, which may not be ideal. Though it is not critical to change it, I feel like it would simplify the flow for Meson.

File: meson.build
Comment (kasper93) [reply meson.build:394]: Thank you for explanation, I will see if this is feasible for me to integrate on each build.

File: compare/image.c
Comment (rcombs) [reply compare/image.c:67]: This has little to do with mingw vs MSVC deps and everything to do with dynamic deps vs static deps. See [this Microsoft docs article](https://learn.microsoft.com/en-us/cpp/c-runtime-library/potential-errors-passing-crt-objects-across-dll-boundaries?view=msvc-170): on Windows, it should be presumed unsafe to pass libc structures across DLL boundaries, or to e.g. `malloc()` an object in one DLL and `free()` it in another. Dynamically-linked deps aren't uncommon by any means, and IMO are a case worth having tests for in our CI.

There are two related potential issues here:
- Different libraries being linked against different *versions* of the libc code (mingw vs MSVC, but also eg MSVC 2013 vs 2015, or UCRT vs MSVCRT)
- Different libraries calling into different *copies* of the same libc code (e.g. either [or both] are statically linked against UCRT)

In either of these cases (which are very normal, esp since statically linking the libc is common on windows, and indeed is enabled by default in some situations), the DLLs have separate heaps, and it is not safe to pass libc/CRT objects (including `FILE*`) across the DLL boundary. This is why we have #775, and why it's so common for libraries to have similar routines (ffmpeg, openssl, etc). This is a common error, and seems extremely worth having tests for!

File: compare/meson.build
Comment (rcombs) [reply compare/meson.build:18]: Meson runs `sh` with `run-all.sh` as the first arg.

File: compare/meson.build
Comment (rcombs) [reply compare/meson.build:18]: I already have a step that prints the full log if tests fail.

Comment (rcombs) [reply compare/meson.build:18]: But anyway, as far as I'm aware, meson doesn't support globbing or other methods of listing the contents of a directory, so this suggestion is not actually possible without hardcoding the list of tests (from `libass/libass-tests`) within `libass/libass`. I suppose it might be possible to `run_command('ls')` and then parse the output? But that'd mean pulling in a list of tests at configure-time, which would be brittle (i.e. if you added a test case, it wouldn't get pulled in until you re-ran meson), and I don't think it'd provide any actual *benefit* here.

File: compare/image.c
Comment (kasper93) [reply compare/image.c:67]: > Different libraries being linked against different versions of the libc code (mingw vs MSVC, but also eg MSVC 2013 vs 2015, or UCRT vs MSVCRT)

That's historically true, and for that reason many binary releases are/were shipping multiple .dll versions built with each of the CRT. But Microsoft resolved those issue with UCRT. They no longer break compatibility on each version.

It's even noted in the article linked

> Because the DLL version of the Universal CRT library used by Visual Studio 2015 and later is now a centrally deployed Windows component (ucrtbase.dll), it's the same for apps built with Visual Studio 2015 and later versions. 

I strongly believe that in present year, there is no need to care about MSVC 2013 (EOL Apr 9, 2019) or MSVC 2015 (EOL Oct 13, 2020) or about MSVCRT at all. Everyone should be using UCRT which is available since Windows Vista SP2 (with an update). And if anyone needs to deploy MSVCRT they should likely built their dependencies to avoid any issues.

MinGW switched by default to UCRT, which in turn is compatible with MSVC builds.

(the only reason MSVCRT is still popular is because prebuilt toolchains in linux distros are outdated)

DLLs on windows are self-contained, unlike in POSIX world symbols are not loaded into global namespaces and by design thy can use different libc or have it duplicated.

> This is a common error, and seems extremely worth having tests for!

Sure. For me this is Frankenstein build, it doesn't reflect real scenario. And if we want to test corner cases, we should also have nominal cases tested, otherwise what is the point?

File: compare/meson.build
Comment (kasper93) [reply compare/meson.build:18]: > But anyway, as far as I'm aware, meson doesn't support globbing or other methods of listing the contents of a directory

And that's a good thing. But you can list dir or whatever you need with a script. https://mesonbuild.com/FAQ.html#but-i-really-want-to-use-wildcards

File: compare/image.c
Comment (astiob) [reply compare/image.c:67]: I didn’t expect that this affects static builds even when using the same libc, but it makes sense.

Anyway, I’d appreciate a short code comment here explaining this is done to ensure the same stdio is called to open the file and to read it regardless of what stdio libpng is linked to. (Technically, this also affects non-Windows; it’s just more rare.)

> MinGW switched by default to UCRT, which in turn is compatible with MSVC builds.

Not sure what this means. MSYS2 has multiple environments with different C libraries (all of them with MinGW-w64 compilers), and you choose one of them explicitly. This is what I suggested changing in the CI configuration.

File: compare/meson.build
Comment (astiob) [reply compare/meson.build:18]: Interesting.

[This doc page](https://mesonbuild.com/External-commands.html) (while not specifically about `test`) says “Meson will autodetect scripts with a shebang line and run them with the executable/interpreter specified in it both on Windows and on Unixes”. I’m gonna assume tests do the same, and that on Windows it discards the explicit path from the shebang and tries to run the executable by basename from PATH etc.

File: compare/image.c
Comment (kasper93) [reply compare/image.c:67]: > Not sure what this means. MSYS2 has multiple environments with different C libraries (all of them with MinGW-w64 compilers), and you choose one of them explicitly. This is what I suggested changing in the CI configuration.

MSYS2 is not MinGW. MSYS2 is package manager. They provide various environments. See, https://www.msys2.org/docs/environments/ especially MSVCRT vs UCRT section.

Anyway, what I meant is that [MinGW](https://www.mingw-w64.org/) starting from v12.0.0 is using UCRT by default. Of course you can build MSVCRT version (like MSYS2 provides), but we are at the point where we should be migrating everything to UCRT.

See also https://sourceforge.net/p/mingw-w64/mingw-w64/ci/master/tree/mingw-w64-doc/howto-build/ucrt-vs-msvcrt.txt

> This is what I suggested changing in the CI configuration.

This is good suggestion, also worth testing is CLANG64 environment.

File: compare/meson.build
Comment (kasper93) [reply compare/meson.build:18]: Also my point is mostly, that if MSVC Windows support is something that is a real target it should work without third-party tools like sh.exe (above other things) from MSYS2. I know it is currently not feasible, but this is generally something to keep in mind.

File: meson.build
Comment (rcombs) [reply meson.build:380]: A couple of these have actual dependency interactions, and it didn't make sense for some of the utils to be features while others were booleans.

File: compare/image.c
Comment (rcombs) [reply compare/image.c:71]: Added comment clarifying intent.

**Review by astiob — APPROVED** [review]

File: .github/workflows/meson.yml
Comment (TheOneric) [inline .github/workflows/meson.yml:185]: noteworthy: like shown here, the option changes affect the build config interface, which should be properly announcement in an eventual release text. But with how the commits are named and not mentioning this in their desription this will now likely be forgotten
Also, can you elaborate on why it was changed?

File: checkasm/meson.build
Comment (TheOneric) [inline checkasm/meson.build:21]: Unlike automake, meson does **not** properly separate compile flags for assembly and C cources (see also `c_args: asm_args` below. That’s why for the main lib non-NASM assembly is split off into an internal static lib, built with `asm_args` while C sources use the project default `c_args`.

As it stands, *all* checkasm sources, even on NASM platforms, do not use the usual C compiler arguments at all iiuc, but only the much smaller and different set meant for non-NASM assembly files (potentially subtly breaking it in some configs but in any event disabling our warnings etc)*

File: meson.build
Comment (TheOneric) [inline meson.build:128]: This is suboptimal, it will also try to look up libpng and show messages about it to the builder even when no enabled build target uses it. Instead the lookup should be dependend on users being enabled like before, and the dep then being required *(the default iirc)*.

File: compare/image.c
Comment (TheOneric) [reply compare/image.c:71]: reminder: `test` writes pngs and so does `compare`. If we consider this a supported usecase this needs fixing too and if not this is change can be removed once we stop mixing in CI

Comment (rcombs) [reply compare/image.c:71]: Yes, that should be added; I considered it out-of-scope for this PR (esp since we don't have any tests that use those, so it'd be hard to validate).

File: .github/workflows/meson.yml
Comment (rcombs) [reply .github/workflows/meson.yml:185]: > A couple of these have actual dependency interactions, and it didn't make sense for some of the utils to be features while others were booleans.

File: compare/image.c
Comment (TheOneric) [reply compare/image.c:71]: > esp since we don't have any tests that use those, so it'd be hard to validate

The muon CI job actually builds and executes the `test` utility atm though it ofc doesn't use MSVC. With the PR here it now makes more sense to run the usual tests instead

File: checkasm/meson.build
Comment (TheOneric) [reply checkasm/meson.build:21]: Does MSVC accept `-D` macro defines? If not this might actually already break for MSVC Windows ARM builds

Comment (rcombs) [reply checkasm/meson.build:21]: MSVC does accept `-D` args.

As far as I can tell, the full set of C args *is* passed here; an example command from a local build:
`[40/49] cc -Icheckasm/checkasm.p -Icheckasm -I../checkasm -I. -I.. -Ilibass -I../libass -I/opt/homebrew/opt/freetype/include/freetype2 -I/opt/homebrew/opt/libpng/include/libpng16 -I/opt/homebrew/Cellar/fribidi/1.0.15/include/fribidi -I/opt/homebrew/Cellar/harfbuzz/9.0.0/include/harfbuzz -I/opt/homebrew/Cellar/glib/2.80.4/include/glib-2.0 -I/opt/homebrew/Cellar/glib/2.80.4/lib/glib-2.0/include -I/opt/homebrew/opt/gettext/include -I/opt/homebrew/Cellar/pcre2/10.44/include -I/opt/homebrew/Cellar/graphite2/1.3.14/include -I/opt/homebrew/Cellar/libunibreak/6.1/include -I/opt/homebrew/Cellar/fontconfig/2.14.2/include -fdiagnostics-color=always -Wall -Winvalid-pch -Wextra -std=gnu11 -O2 -g -Wno-sign-compare -Wno-unused-parameter -Werror-implicit-function-declaration -Wstrict-prototypes -Wpointer-arith -Wredundant-decls -Wno-missing-field-initializers -Wformat-non-iso -Werror=format-non-iso -Watomic-implicit-seq-cst -D_GNU_SOURCE -D_XPLATFORM_SOURCE -fno-math-errno -DPREFIX -MD -MQ checkasm/checkasm.p/checkasm.c.o -MF checkasm/checkasm.p/checkasm.c.o.d -o checkasm/checkasm.p/checkasm.c.o -c ../checkasm/checkasm.c`

So it gets everything we passed to `add_project_arguments`, plus `asm_args` (which is just `-DPREFIX`, which is harmless to apply to our C sources as well).

That said, we should really add a ARM builds to the CI (both on Windows and Linux).

File: meson.build
Comment (rcombs) [reply meson.build:128]: Does this really matter? It's a one-time cached configure check (and tbh we might end up turning the test/compare/etc programs on by default behind a libpng-availability gate at some point anyway).

**Pull request merged**
